src/quicktype-core/Constraints.ts

```diff
@@ -0,0 +1,186 @@
+        return undefined;
+    }
+
+    stringify([min, max]: MinMaxConstraint): string {
+        return `${min}-${max}`;
+    }
+}
+
+export const minMaxTypeAttributeKind: TypeAttributeKind<MinMaxConstraint> = new MinMaxConstraintTypeAttributeKind(
+    "minMax"
+);
+
+export const minMaxLengthTypeAttributeKind: TypeAttributeKind<MinMaxConstraint> = new MinMaxConstraintTypeAttributeKind(
+    "minMaxLength"
+);
+
+function producer(
+    schema: JSONSchema,
+    types: Set<JSONSchemaType>,
+    minProperty: string,
+    maxProperty: string
+): MinMaxConstraint | undefined {
+    if (!(typeof schema === "object")) return undefined;
+    if (!types.has("number") && !types.has("integer") && !types.has("string")) return undefined;
+
+    let min: number | undefined = undefined;
+    let max: number | undefined = undefined;
+
+    if (typeof schema[minProperty] === "number") {
+        min = schema[minProperty];
+    }
+    if (typeof schema[maxProperty] === "number") {
+        max = schema[maxProperty];
+    }
+
+    if (min === undefined && max === undefined) return undefined;
+    return [min, max];
+}
+
+export function minMaxAttributeProducer(
+    schema: JSONSchema,
+    _ref: Ref,
+    types: Set<JSONSchemaType>
+): JSONSchemaAttributes | undefined {
+    const maybeMinMax = producer(schema, types, "minimum", "maximum");
+    if (maybeMinMax === undefined) return undefined;
+    return { forNumber: minMaxTypeAttributeKind.makeAttributes(maybeMinMax) };
+}
+
+export function minMaxLengthAttributeProducer(
+    schema: JSONSchema,
+    _ref: Ref,
+    types: Set<JSONSchemaType>
+): JSONSchemaAttributes | undefined {
+    const maybeMinMaxLength = producer(schema, types, "minLength", "maxLength");
+    if (maybeMinMaxLength === undefined) return undefined;
+    return { forString: minMaxLengthTypeAttributeKind.makeAttributes(maybeMinMaxLength) };
+}
+
+export function minMaxValue(t: Type): MinMaxConstraint | undefined {
+    return minMaxTypeAttributeKind.tryGetInAttributes(t.getAttributes());
+}
+
+export function minMaxLength(t: Type): MinMaxConstraint | undefined {
+    return minMaxLengthTypeAttributeKind.tryGetInAttributes(t.getAttributes());
+}
+
+export class PatternTypeAttributeKind extends TypeAttributeKind<string> {
+    get inIdentity(): boolean {
+        return true;
+    }
+
+    combine(arr: string[]): string {
+        assert(arr.length > 0);
+
+        let patt:string="";
+        for (let i = 1; i < arr.length; i++) {
+            patt += "("+arr[i]+")";
+            if (i !== arr.length-1) {
+                patt += "|";
+            }
+        }
+        return patt;
+    }
+
+    intersect(_arr: string[]): string | undefined {
+        /** FIXME!!! what is the intersection of regexps? */
+        return undefined;
+    }
+
+    makeInferred(_: string): undefined {
+        return undefined;
+    }
+}
+
+export const patternTypeAttributeKind: TypeAttributeKind<string> = new PatternTypeAttributeKind(
+    "pattern"
```

Comment:
I think the `PatternTypeAttributeKind` should have a constructor that calls `super("pattern")`.  We only need one.

Reply:
Still not resolved.